### ISSUE TITLE
Bug: [prefer-readonly] False positive if property was assigned after type guard for class properties

### ISSUE THREAD

**Issue by kutaxoff** [typescript-eslint/typescript-eslint#7221]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.1.6&fileType=.tsx&code=MYGwhgzhAEAqCmEAu0DeAoa0AOBXARiAJbDQB288AJhLAPYBC8AIvAGZEVUD8AXObgC2%2BeACcA3Okw5RRAG5gk8cnSQAleGCp0yIAJ4AxOnT4DhYydOyyFSles3bdehmFGmyQkRKlY8hEmglZABZeCQACzoqIzoACgBKNGksIjZoOIBCSKIIADpQSAgABVE6bDEkPQBxXDcqRKSMLBboUXDcUTJJVoBfFKCI3LyyVQ0tHX1Y6ABeaABGHuh%2BqwJiUmCkMMjo11FE5Na0jOyh-NHRQTAQWD0K2vq4nPPKGnomVg4uBKaBrHakJ1ugN%2Bq1niMxo5Ji43LNoAAmJYrPw2RTKQpQUrlSo1OqiBoJfjPaC5QakgBkaHIr1ojBY7E41H4nnMomWhxaAKBZJe1FpHwZXGgmRmc1wZCoguoSN8MnkaJUl2ut3ueIaChAuHgzK8YmgAB9oOLJV9qIToBqtSSYCzvBz-h0uhbrlaRWKJVKqDLekA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wls4CGAEwD2TeIXRRe0EdEjgwAXxBKgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylAGYBDVCXTgwAXxCygA&tokens=false

### Repro Code

```typescript
class Test {
  public needsToBeDefined?: number;

  private notReadonlyFoo?: number;

  private notReadonlyBar?: number;

  public testMethodFoo() {
    if (!this.classPropertyGuard()) {
      return;
    }
    this.notReadonlyFoo = 1;
  }

  public testMethodBar() {
    if (!this.normalTypeGuard(this.needsToBeDefined)) {
      return;
    }
    this.notReadonlyBar = 2;
  }

  private classPropertyGuard(): this is this & { needsToBeDefined: number } {
    return this.needsToBeDefined !== undefined;
  }

  private normalTypeGuard(value: number | undefined): value is number {
    return value !== undefined;
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/prefer-readonly": "error"
  },
};
```

### Expected Result

I expected that `private notReadonlyFoo?: number;` would not report any errors.

### Actual Result

`@typescript-eslint/prefer-readonly` error on the 4th line: 
```
Member 'notReadonlyFoo' is never reassigned; mark it as `readonly`. 4:3 - 4:35
```

**Comment by kutaxoff** [comment]
`notReadonlyFoo` is clearly reassigned in the `testMethodFoo()` method, but eslint instructs to mark it as readonly. It is possible to reach the code with assignment after type guard. If you use classic type guard like `normalTypeGuard(value)` there is no error, but it is present if you use a hack to typeguard class member ([like here](https://github.com/microsoft/TypeScript/issues/26212#issuecomment-410536871))

**Comment by JoshuaKGoldberg** [comment]
Ah, got it - thanks. That makes more sense. This does seem like a bug. 

I found the original code snippets a little confusing so I made a smaller reproduction case: https://typescript-eslint.io/play/#ts=5.1.6&fileType=.tsx&code=MYGwhgzhAEAqCmEAu0DeAoa0AOBXARiAJbDQC2YAnvvAPwBc0AdrmTQE4Dc6mO7RANzBJ4zAPZIASvDAATMUxCUAcmPYUQDZqw7de2fkJHipM%2BYsqwAFkQhaWbeF14B6F9ACy8R%2B2gByJglpOQUla1s-aFtmeAEnaHYZKCIAcyZ4WU5yMHYAayiUSGgAA0SQi2KAOmgeLDxCEmgRZFhKbHgAcVwc2QgACgBKNF4sIgAzaD6kGwhK4Ct4YFzVdTAQQaGMLG2mmcrA03KlFY1oAF5oAAZuHYBfWu3xyenbOYWl1vaunvD%2Bgc2RtsXrMDsFzGEZucrjdtvcsPd9IZhKJ5otlmoNINhjtEkhcOwmLtXhRqKIAIRnC64JiyeBjIjpTK8BF1JHGVEfNqdbrsWS-QaMYFRGBCgBkaGypMYDg40Fu2O2uPxhOBlRJNGgFKpNLpDIyMPutyAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wls4CGAEwD2TeIXRRe0EdEjgwAXxBKgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylAGYBDVCXTgwAXxCygA&tokens=false

```ts
class Test {
  public maybe?: number;

  private notReadonlyNormal?: number;

  private notReadonlyThis?: number;
  // Member 'notReadonlyThis' is never reassigned; mark it as `readonly`. 

  public testTypeGuards() {
    if (this.checkNormal()) {
      this.notReadonlyNormal = 0;
    }

    if (this.checkTypeGuardThis()) {
      this.notReadonlyThis = 0;
    }
  }

  private checkNormal() {
    return this.maybe !== undefined;
  }

  private checkTypeGuardThis(): this is this & { maybe: number } {
    return this.maybe !== undefined;
  }
}
```
